File: backend/app/Models/Bike.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

#[Fillable([
    'code',
    'name',
    'status',
    'current_latitude',
    'current_longitude',
    'last_accuracy',
    'is_online',
    'battery_percent',
    'assigned_device_user_id',
    'last_seen_at',
])]
class Bike extends Model
{
    use HasFactory;

    protected function casts(): array
    {
        return [
            'current_latitude' => 'decimal:7',
            'current_longitude' => 'decimal:7',
            'last_accuracy' => 'decimal:2',
            'is_online' => 'boolean',
            'last_seen_at' => 'datetime',
        ];
    }

    public function assignedDevice(): BelongsTo
    {
        return $this->belongsTo(User::class, 'assigned_device_user_id');
    }

    public function rentals(): HasMany
    {
        return $this->hasMany(Rental::class);
    }

    public function activeRental(): HasOne
    {
        return $this->hasOne(Rental::class)
            ->whereIn('status', [
                Rental::STATUS_ACTIVE,
                Rental::STATUS_IDLE_WARNING,
                Rental::STATUS_IDLE_BILLING,
            ])
            ->latestOfMany('started_at');
    }

    public function locationPoints(): HasMany
    {
        return $this->hasMany(RentalLocationPoint::class);
    }

    public function latestLocationPoint(): HasOne
    {
        return $this->hasOne(RentalLocationPoint::class)->latestOfMany('recorded_at');
    }

    public function deviceHeartbeats(): HasMany
    {
        return $this->hasMany(DeviceHeartbeat::class);
    }

    public function cellObservations(): HasMany
    {
        return $this->hasMany(CellObservation::class);
    }

    public function latestHeartbeat(): HasOne
    {
        return $this->hasOne(DeviceHeartbeat::class)->latestOfMany('last_seen_at');
    }
}

File: mobile_bike/lib/src/models/bike.dart
class Bike {
  Bike({
    required this.id,
    required this.code,
    required this.name,
    required this.status,
    required this.isOnline,
    this.batteryPercent,
    this.currentLatitude,
    this.currentLongitude,
    this.lastAccuracy,
    this.lastSeenAt,
  });

  factory Bike.fromJson(Map<String, dynamic> json) {
    return Bike(
      id: json['id'] as int,
      code: json['code'] as String,
      name: json['name'] as String,
      status: json['status'] as String,
      isOnline: (json['is_online'] as bool?) ?? false,
      batteryPercent: json['battery_percent'] as int?,
      currentLatitude: json['current_latitude'] != null
          ? double.tryParse(json['current_latitude'].toString())
          : null,
      currentLongitude: json['current_longitude'] != null
          ? double.tryParse(json['current_longitude'].toString())
          : null,
      lastAccuracy: json['last_accuracy'] != null
          ? double.tryParse(json['last_accuracy'].toString())
          : null,
      lastSeenAt: json['last_seen_at'] != null
          ? DateTime.tryParse(json['last_seen_at'].toString())?.toLocal()
          : null,
    );
  }

  final int id;
  final String code;
  final String name;
  final String status;
  final bool isOnline;
  final int? batteryPercent;
  final double? currentLatitude;
  final double? currentLongitude;
  final double? lastAccuracy;
  final DateTime? lastSeenAt;
}
